heimdall: Astronomy software package

aweaver1fandm opened the pull request:
Heimdall requires two additional astronomy software packages, DEDISP and PSRDADA.  Package files for these have also been included.

aweaver1fandm committed:
heimdall: Astronomy software package.  Requires dedisp and psrdada (included as part of this commit)

aweaver1fandm committed:
Merge branch 'develop' into packages/heimdall

wdconinc commented on var/spack/repos/builtin/packages/dedisp/package.py:
```diff
@@ -0,0 +1,40 @@
+
+        # The $PREFIX/dedisp/include and $PREFIX/dedisp/lib directories don't seem
+        # to be created automatically by the software's Makefile so manually create them
+        libdir = os.path.join(prefix, 'lib')
+        incdir = os.path.join(prefix, "include")
+        os.mkdir(libdir)
+        os.mkdir(incdir)
```
Without need for `os` and using already imported functionality:
```suggestion
        libdir = join_path(prefix, 'lib')
        incdir = join_path(prefix, "include")
        mkdirp(libdir)
        mkdirp(incdir)
```
Refs: https://spack.readthedocs.io/en/latest/llnl.util.html#llnl.util.filesystem.join_path, https://spack.readthedocs.io/en/latest/llnl.util.html#llnl.util.filesystem.mkdirp

wdconinc commented on var/spack/repos/builtin/packages/dedisp/package.py:
```diff
@@ -0,0 +1,40 @@
+    homepage = "https://github.com/ajameson/dedisp"
+
+    git = "https://github.com/ajameson/dedisp.git"
+    version("1.0", branch="master")
```
I think it is better in this case to list this as version `master` to avoid future confusion.
```suggestion
    version("master", branch="master", preferred=True)
```
See e.g. `flexi`. https://github.com/spack/spack/blob/1ac2ee8043ab802ff5c9161ac35c269f22df3c34/var/spack/repos/builtin/packages/flexi/package.py#L18

You'll see in `flexi` that you can also use specific commits with a calendar version for additional stability in versions.

wdconinc commented on var/spack/repos/builtin/packages/heimdall/package.py:
```diff
@@ -0,0 +1,47 @@
+    homepage = "https://sourceforge.net/projects/heimdall-astro/"
+    git = "https://git.code.sf.net/p/heimdall-astro/code"
+
+    version("1.0", branch="master")
```
```suggestion
    version("master", branch="master", preferred=True)
```

wdconinc commented on var/spack/repos/builtin/packages/heimdall/package.py:
```diff
@@ -0,0 +1,47 @@
+
+    # Pass the cuda architecture to DEDISP and PSRDADA for building
+    for arch in CudaPackage.cuda_arch_values:
+        depends_on("dedisp +cuda cuda_arch=%s" % arch,
+                   when="+cuda cuda_arch=%s" % arch)
+
+        depends_on("psrdada +cuda cuda_arch=%s" % arch,
+                   when="+cuda cuda_arch=%s" % arch)
```
We prefer f-strings now.
```suggestion
        depends_on(f"dedisp +cuda cuda_arch={arch}",
                   when=f"+cuda cuda_arch={arch}")

        depends_on(f"psrdada +cuda cuda_arch={arch}",
                   when=f"+cuda cuda_arch={arch}")
```

wdconinc commented on var/spack/repos/builtin/packages/heimdall/package.py:
```diff
@@ -0,0 +1,47 @@
+
+    def configure_args(self):
+        # Heimdall requires some special flags for configure
+        args = ["--with-psrdada-dir=" + self.spec["psrdada"].prefix,
+                "--with-dedisp-dir=" + self.spec["dedisp"].prefix,
+                "--with-cuda-dir=" + self.spec["cuda"].prefix]
```
```suggestion
        args = [f"--with-psrdada-dir={self.spec['psrdada'].prefix}",
                f"--with-dedisp-dir={self.spec['dedisp"].prefix}",
                f"--with-cuda-dir={self.spec['cuda'].prefix}"]
```

wdconinc commented on var/spack/repos/builtin/packages/heimdall/package.py:
```diff
@@ -0,0 +1,47 @@
+                "--with-cuda-dir=" + self.spec["cuda"].prefix]
+        return args
+
+    def install(self, spec, prefix):
+        make()
+        make("install")
```
Since you do not modify the default behavior of `install`, you can omit this.
```suggestion
```

wdconinc commented on var/spack/repos/builtin/packages/psrdada/package.py:
```diff
@@ -0,0 +1,28 @@
+    depends_on("libtool", type="build")
+    depends_on("python")
+    depends_on("cuda@:11.7", type="build")
+
+    def install(self, spec, prefix):
+        make()
+        make("install")
```
```suggestion
```

wdconinc reviewed (commented):
Thanks for your contribution! Apologies for the delay in reviewing this. I have added some comments below.

Please also rebase or merge updates to ensure the latest audits and style checks are applied.

aweaver1fandm committed:
Updated packages to align with Spack's style

Minor updates based on wdconinc's comments regarding Spack's style guide

aweaver1fandm committed:
Merge branch 'develop' into packages/heimdall

wdconinc commented on var/spack/repos/builtin/packages/dedisp/package.py:
```diff
@@ -0,0 +1,42 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

wdconinc commented on var/spack/repos/builtin/packages/heimdall/package.py:
```diff
@@ -0,0 +1,46 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

wdconinc commented on var/spack/repos/builtin/packages/psrdada/package.py:
```diff
@@ -0,0 +1,30 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

wdconinc commented on var/spack/repos/builtin/packages/dedisp/package.py:
```diff
@@ -0,0 +1,42 @@
+        makefile.filter(r'^\s*GPU_ARCH\s*\?=.*', 'GPU_ARCH ?= sm_' + spec.variants['cuda_arch'].value[0])
+        makefile.filter(r'^\s*INSTALL_DIR\s*\?=.*', 'INSTALL_DIR ?= ' + prefix)
+
+    def install(self, spec, prefix):
+
+        # The $PREFIX/dedisp/include and $PREFIX/dedisp/lib directories don't seem
+        # to be created automatically by the software's Makefile so manually create them
+        libdir = join_path(prefix, 'lib')
+        incdir = join_path(prefix, 'include')
+        mkdirp(libdir)
+        mkdirp(incdir)
+
+        make()
+        make("install")
```
I was thinking more about this, and it is possible that you can replace this with:
```suggestion
    @run_before("install"):
    def preinstall(self, spec, prefix):

        # The $PREFIX/dedisp/include and $PREFIX/dedisp/lib directories don't seem
        # to be created automatically by the software's Makefile so manually create them
        libdir = join_path(prefix, 'lib')
        incdir = join_path(prefix, 'include')
        mkdirp(libdir)
        mkdirp(incdir)
```
This will reuse the default `install` function instead of reimplementing it, but run the function before it.

wdconinc reviewed (commented):
Couple more comments (three minor about date in copyright). Can you copy/paste an install test build summary? I don't have a CUDA install to verify right now.

wdconinc commented:
FYI, for the style checks, you can apply these automatically (most of the time) with the following bot command (not sure it will work as part of this command, but if it doesn't, feel free to repeat as standalone PR comment):

@spackbot fix style

aweaver1fandm committed:
[@spackbot] updating style on behalf of aweaver1fandm

aweaver1fandm committed:
Minor edits to fix copyright year and dedisp install

Fixed copyright year to be 2024 instead of 2023

Removed the overridden version of install and created a preinstall function to create the missing lib and inc directories, therefore allowing the default install to run

Here is output from the spack install of heimdall showing successful build with cuda.  If necessary I do a spack clean and freshly install it
./spack install heimdall +cuda cuda_arch=86
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/libiconv-1.17-enpmbhsi3kztebwmpclpub2afhlbr3gy
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/xz-5.4.1-pte76kujkezxb3laqse3o4sctlbygsaw
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/zlib-1.2.13-utlfo5ltxz5v5bckirn5v3amtbxjdvwh
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/bzip2-1.0.8-x6navz7ucgfnb5xq7aelqmgd4zxsz5bs
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/libmd-1.0.4-ncomhrodpdul4dm64o6b7426fhmc2u64
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/ncurses-6.4-c77h34rooycbzapxjvc27sg5td5jiwyb
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/zstd-1.5.5-eporpybumydxveg5rwtfzysrsu4eqzcv
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/libffi-3.4.4-vskntokaojclfqxjfzkbyirkeogddbpx
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/libxcrypt-4.4.33-gtpn32p6mxztul3c3dxzqj7gvcyh555j
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/util-linux-uuid-2.38.1-g322a5peqjaad6gl5q64cdu4qo7kvw6o
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/libxml2-2.10.3-pza3kz2mtbncbbeim6rejfqkgftnf4rz
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/openssl-1.1.1t-adbquvgg4qpc3vq6jynf44qzq3gfwrv5
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/pigz-2.7-e7mxj4ya2u4a6zb4hu64g7docujmkxeb
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/libbsd-0.11.7-wh4xleivbe7wndiqt5nsehzlfrccnjcg
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/readline-8.2-hhb647bwmbcj7iwpmtetbylninfm5rxf
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/cuda-11.7.1-osue2sx5rv7dgzhsmaemydpwhyribxng
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/tar-1.34-35f5gki2ycxmy5zd7zs5tsvp3xoszxum
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/expat-2.5.0-4gizyhhqklciuyrbyinq2tdggt73gds4
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/gdbm-1.23-w3uzihtubj2iwv6es55fis6nt2q5zwlr
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/sqlite-3.40.1-m4ntzvuupsnbtkdfbz7oqpbjdlaffp2a
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/dedisp-1.0-pq6r3jnyxq6jzoygz3fp2e6jc2ojpvap
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/gettext-0.21.1-hglzdeadmgkzjb76bmemt6dnulfkrpha
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/python-3.10.10-khu36qq4p2te7jf475ewr2h7egidekfl
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/psrdada-master-by4w6mrpcfnoihlhos2jcfo2roiyagaz
[+] /home/htony/spack/opt/spack/linux-ubuntu22.04-zen3/gcc-11.3.0/heimdall-1.0-ohtdnltuhhejysshcert25h6nmuvluqp

wdconinc commented:
@spackbot fix style

aweaver1fandm committed:
[@spackbot] updating style on behalf of aweaver1fandm

wdconinc reviewed (approved):
Looks good! Thanks for the contribution, and for the test result in 79bd4b82434858252caa7f9358a6165d7f9c7e38.

aweaver1fandm committed:
Merge branch 'develop' into packages/heimdall

wdconinc commented:
FYI, updating this branch to too new of a base branch often slows down the merge; see https://github.com/spack/spack-infrastructure/blob/main/docs/deferred_pipelines.md.

aweaver1fandm commented:
Thanks for that bit of info.  I won't update it then

On Mon, May 20, 2024 at 1:29 PM Wouter Deconinck ***@***.***>
wrote:

> FYI, updating this branch to too new of a base branch often slows down the
> merge; see
> https://github.com/spack/spack-infrastructure/blob/main/docs/deferred_pipelines.md
> .
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/38328#issuecomment-2120887574>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANJXBC7LQAKGIDW42H6L45TZDIXGLAVCNFSM6AAAAAAZDLVLAGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCMRQHA4DONJXGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
